README.md
automaton
==========

A simple cellular automata simulation runner in C
(such as Conway's Game of Life)
![automaton](http://i.imgur.com/LMw8CsK.gif)
![briansbraingraphical](http://i.imgur.com/OEZeJu6.png)

Quick Start
-----------
For 64-bit Ubuntu Linux, a tarball with precompiled binary and required assets
is provided for convenience in [automaton/releases](https://github.com/calebsmith/automaton/releases)

Notes
-----

This project is largely experimental and exists for my own learning, but has
recently become pretty stable nonetheless.

This was developed and tested on Ubuntu and Arch GNU/Linux. The curses and time
libraries rely on POSIX standards and will likely compile and run on Mac OSX
with some work, but this is not yet tested at all. (It is my suspicion that the
largest hurdles will be compiling the GLFW3 and Guile 2.0 dependencies).

Windows support is non-existent at the moment. If this is desired, please
contribute by making a pull request or opening an issue with links to good
documentation.

Dependencies
------------

 * ncurses 5.9.6
 * opengl
 * glfw 3.0.4
 * guile 2.0

For Arch these are provided in the following packages:

 * glfw
 * ncurses
 * guile

For Debian/Ubuntu GNU/Linux these are provided in the following packages:

 * libncurses5-dev
 * libglu1-mesa-dev
 * guile-2.0-dev
 * (must compile and install glfw 3 from source)

The base requirements for compiling are:

 * git 1.7.9
 * make 3.8.1
 * gcc 4.6.3

Redistribution
--------------

A `create_build.sh` shell script is available for making redistribution of the
program simpler. It is used to create the tarballs provided on the GitHub
releases page.

Currently, precompiled releases are provided for 64-bit Ubuntu only. Any help
with packaging on other distros is greatly appreciated.

At the moment, the release builds are not static linked. Any help in this
regard would be greatly appreciated.


Installation
------------
If you are on 64-bit Ubuntu, download a release from
[automaton/releases](https://github.com/calebsmith/automaton/releases)
You will likely need libguile-2.0, libncurses5 and libglu1-mesa-dev for
dynamic linking to work.

Otherwise, you'll need to build from source, at least for the moment. First,
ensure all requirements are met.

To build, clone the repo, run make in the src folder, then run any of the
shell scripts beginning with 'run', such as 'run_life.sh' to test it::


    git clone git@github.com:calebsmith/automaton.git
    cd automaton/src
    make
    cd ..
    source run_life.sh


Usage
-----

The program requires one argument, which is the data file for loading the
initial state (e.g. seed). Such files are provided with the repository in
the /data folder. The run_life.sh shell command demonstrates running
automaton with the gosper_gun data file.

The command format is:

automaton [Options] seed_filename

Command line flags include:

* -g, --graphical - Use graphical mode (OpenGL rendering)
* -f, --fullscreen - Use fullscreen rendering rather than windowed (pair with
        -g option, otherwise has no effect)
* -t, --toroidal - Make the board toroidal (wrap around)
* -s number, --speed=number - Set the amount of microseconds between
    generations. (Default is 85000)

Examples

    #Fast, full-screen graphical display
    ./automaton -g -f -s 1000 data/gosper_gun.dat
    #Slow, toroidal textual display
    ./automaton -s 500000 data/r_pentomino.dat


Usage During Runtime
--------------------

During program execution, the following keys have the corresponding effects:

* escape or q - quit
* p - pause
* r - resume
* h / left-arrow - move view left
* j / down-arrow - move view down
* k / up-arrow - move view up
* l / right-arrow - move view right
* i - zoom in
* o - zoom out


To define your own automata and seeds refer to the docs on file formats and the Scheme API:

[File formats](docs/file_formats.md).

[Scheme](docs/scheme_api.md).


Authors
-------
Caleb Smith


docs/change_log.md
Change log
==========

v0.0.9 (Oct 12 2014)
-------------------
    * Changed C/Scheme layer to call Scheme once per generation, not per cell.
        This makes the given module "impure" (it sets values on the board
        directly) but allows for more flexibility in what a given module can
        implement.
    * Added an implementation of Langton's Ant in Scheme.
    * Changed project name to "automaton" from "gameoflife" since GOL is one
        of several automata in the project.

v0.0.8 (June 8 2014)
-------------------
    * Added documentation for file formats and Scheme API
    * Added support for defining automata rules in Scheme.
    * Uses pkg-config for glfw3 linker flags to make builds on different
        distributions simpler

v0.0.7 (May 11 2014)
--------------------
    * Upgraded glfw to version 3
    * Added a "world" module to simplify loading and passing boards and rules
        data
    * Improved Makefile to use object files
    * Fixed a bug to make the error when no filename is given print correctly
    * Fixed some bad calls to free() on program exit under some conditions
        (was crashing on exit for some rulesets)

v0.0.6 (Oct 18 2013)
--------------------
    * Removed hard coded Game of Life algorithm and replaced with a general
        "rule" file to define the automata to use.
    * Added some example data files for WireWorld automata.
    * Added interactive zooming to the lens feature.

v0.0.5 (Oct 13 2013)
--------------------
    * Added a "lens" to handle the display of the board. Allows for boards
        larger than the display surface, and handles logic for zooming,
        navigation, centering and so on.
    * Added navigation of the "lens" across the board using hjkl (ala vim) and
        arrow keys
    * Changed file format of seed data files to allow for easier creation of
        large sparse boards

v0.0.4 (Oct 06 2013)
--------------------
    * Added an OpenGL front-end.
    * Added key handling for pausing, resuming and exiting the simulation
    * Improved exit codes and error reporting
    * Split program into main, board, backend, textual_ui, and graphical_ui
        source files

v0.0.3
------
    * Made non-toroidal board extend beyond the visible bounds, allowing e.g.
        for escaping gliders to leave the visible board before colliding with
        the boundary of the board in memory
    * Several small code style improvements.
    * Improved text display (made cursor invisible, replaced 0 with o.)
    * Made toroidal board and speed command line options rather than loading
        these from data files.

v0.0.2
------
    * Added curses requirement and made text display consistent
    * Fixed CPU waiting to use POSIX nanosleep rather than the deprecated
        usleep.

v0.0.1
------
    * Initial public release



src/main.c
/*
A cellular automata runner in C, with extensibility in Scheme
Copyright (C) 2013  Caleb Smith

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#include "world.h"
#include "scm.h"
#include "text_ui.h"
#include "graphical_ui.h"


const char* USAGE_STRING = ""
"The command format is:\n"
"\n"
"automaton [Options] seed_filename\n"
"\n"
"Command line flags include:\n"
"\n"
"* -g, --graphical - Use graphical mode (OpenGL rendering)\n"
"* -f, --fullscreen - Use fullscreen rendering rather than windowed (pair with\n"
"        -g option, otherwise has no effect)\n"
"* -t, --toroidal - Make the board toroidal (wrap around)\n"
"* -s number, --speed=number - Set the amount of microseconds between\n"
"    generations. (Default is 85000)\n"
"\n"
"Examples\n"
"\n"
"    #Fast, full-screen graphical display\n"
"    ./automaton -g -f -s 1000 data/gosper_gun.dat\n"
"    #Slow, toroidal textual display\n"
"    ./automaton -s 500000 data/r_pentomino.dat\n"
"\n"
"\n"
"Usage During Runtime\n"
"--------------------\n"
"\n"
"During program execution, the following keys have the corresponding effects:\n"
"\n"
"* escape or q - quit\n"
"* p - pause\n"
"* r - resume\n"
"* h / left-arrow - move view left\n"
"* j / down-arrow - move view down\n"
"* k / up-arrow - move view up\n"
"* l / right-arrow - move view right\n"
"* i - zoom in\n"
"* o - zoom out\n";


World_t world_g;

void inner_main(void *nop, int argc, char** argv);


int main(int argc, char* argv[])
{
    scm_boot_guile(argc, argv, &inner_main, NULL);
    return 0;
}


void inner_main(void *nop, int argc, char** argv)
{
    Config_t config;

    // Parse the command line arguments and store into config
    config = get_config(argc, argv);
    if (config.help) {
        printf("%s", USAGE_STRING);
        return;
    }
    // Load board and ruleset into 'world'
    if (world_init(&world_g, config)) {
        return;
    }
    // Register C primitive functions for Scheme
    scm_with_guile(&register_scm_functions, NULL);
    // Enter main loop
    if (config.graphical) {
        main_glfw(&world_g, config.sleep_time * 1000, config.fullscreen);
    } else {
        main_curses(&world_g, config.sleep_time * 1000);
    }
    world_destroy(&world_g);
}
